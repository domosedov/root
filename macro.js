/* eslint-disable no-undef */
const { createMacro } = require('babel-plugin-macros');
const babelPlugin = require('effector/babel-plugin');

module.exports = createMacro(logger, {
  configName: 'effectorRoot',
});

function logger({
  references,
  state,
  babel,
  config: { importModuleName = 'effector-root', ...config } = {},
}) {
  const program = state.file.path;
  // First of all replace import path to effector-root instead of /macro
  Object.keys(references).forEach((referenceName) => {
    const id = addImport(babel.types, program, referenceName, importModuleName);

    // Change name of method to updated
    references[referenceName].forEach((referencePath) => {
      referencePath.node.name = id;
    });
  });

  const instance = babelPlugin(babel, config);

  instance.pre();
  babel.traverse(program.parent, instance.visitor, undefined, {
    ...state,
    ...instance,
  });
  instance.post();
}

function addImport(t, programPath, specifierName, importPath) {
  const [newPath] = programPath.unshiftContainer(
    'body',
    t.importDeclaration(
      [
        t.importSpecifier(
          programPath.scope.generateUidIdentifier(specifierName),
          t.identifier(specifierName),
        ),
      ],
      t.stringLiteral(importPath),
    ),
  );

  let found;

  newPath.get('specifiers').forEach((specifier) => {
    if (specifier.node.imported.name === specifierName) {
      found = specifier;
    }
  });

  programPath.scope.registerBinding('module', found);
  return found.node.local.name;
}
