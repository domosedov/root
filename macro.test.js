/* eslint-disable no-undef */
const { default: pluginTester } = require('babel-plugin-tester');

const baseCode = `
import { createStore, createEvent, createEffect, attach, restore, root } from './macro';
const fx = createEffect(() => 1);
const $data = createStore(0);
const was = createEvent();
const demo = root.createEvent();
const anotherFx = attach({
  effect: fx,
  source: $data,
  mapParams: (a) => a,
})
const $has = restore(was, "");
`;

pluginTester({
  pluginName: 'effector-root/macro',
  plugin: require('babel-plugin-macros'),
  root: __dirname,
  filename: __filename,
  babelOptions: { filename: __filename },
  snapshot: true,
  tests: {
    'add sid': {
      code: baseCode,
    },
    'add sid and loc': {
      code: baseCode,
      pluginOptions: {
        effectorRoot: {
          addLoc: true,
          addNames: true,
        },
      },
    },
  },
});
