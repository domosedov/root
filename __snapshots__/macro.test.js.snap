// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`effector-root/macro add sid and loc: add sid and loc 1`] = `

import { createStore, createEvent, createEffect, attach, restore, root } from './macro';
const fx = createEffect(() => 1);
const $data = createStore(0);
const was = createEvent();
const demo = root.createEvent();
const anotherFx = attach({
  effect: fx,
  source: $data,
  mapParams: (a) => a,
})
const $has = restore(was, "");

      ↓ ↓ ↓ ↓ ↓ ↓

var _effectorFileName = '/macro.test.js';
import { root as _root } from 'effector-root';
import { restore as _restore } from 'effector-root';
import { attach as _attach } from 'effector-root';
import { createEffect as _createEffect } from 'effector-root';
import { createEvent as _createEvent } from 'effector-root';
import { createStore as _createStore } from 'effector-root';

const fx = _createEffect(() => 1, {
  loc: {
    file: _effectorFileName,
    line: 2,
    column: 11,
  },
  name: 'fx',
  sid: '-u6p3sc',
});

const $data = _createStore(0, {
  loc: {
    file: _effectorFileName,
    line: 3,
    column: 14,
  },
  name: '$data',
  sid: '8tv02c',
});

const was = _createEvent({
  loc: {
    file: _effectorFileName,
    line: 4,
    column: 12,
  },
  name: 'was',
  sid: '-3z5gxy',
});

const demo = _root.createEvent({
  loc: {
    file: _effectorFileName,
    line: 5,
    column: 13,
  },
  name: 'demo',
  sid: 'xaneri',
});

const anotherFx = _attach({
  ɔ: {
    effect: fx,
    source: $data,
    mapParams: (a) => a,
  },
  config: {
    loc: {
      file: _effectorFileName,
      line: 6,
      column: 18,
    },
    name: 'anotherFx',
    sid: '-3vcg62',
  },
});

const $has = _restore(was, '', {
  loc: {
    file: _effectorFileName,
    line: 11,
    column: 13,
  },
  name: '$has',
  sid: 'd9tvji',
});


`;

exports[`effector-root/macro add sid: add sid 1`] = `

import { createStore, createEvent, createEffect, attach, restore, root } from './macro';
const fx = createEffect(() => 1);
const $data = createStore(0);
const was = createEvent();
const demo = root.createEvent();
const anotherFx = attach({
  effect: fx,
  source: $data,
  mapParams: (a) => a,
})
const $has = restore(was, "");

      ↓ ↓ ↓ ↓ ↓ ↓

import { root as _root } from 'effector-root';
import { restore as _restore } from 'effector-root';
import { attach as _attach } from 'effector-root';
import { createEffect as _createEffect } from 'effector-root';
import { createEvent as _createEvent } from 'effector-root';
import { createStore as _createStore } from 'effector-root';

const fx = _createEffect(() => 1, {
  name: 'fx',
  sid: '-u6p3sc',
});

const $data = _createStore(0, {
  name: '$data',
  sid: '8tv02c',
});

const was = _createEvent({
  name: 'was',
  sid: '-3z5gxy',
});

const demo = _root.createEvent({
  name: 'demo',
  sid: 'xaneri',
});

const anotherFx = _attach({
  ɔ: {
    effect: fx,
    source: $data,
    mapParams: (a) => a,
  },
  config: {
    name: 'anotherFx',
    sid: '-3vcg62',
  },
});

const $has = _restore(was, '', {
  name: '$has',
  sid: 'd9tvji',
});


`;
